staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -59,3 +59,54 @@ type InTreePlugin interface {
 	// RepairVolumeHandle generates a correct volume handle based on node ID information.
 	RepairVolumeHandle(volumeHandle, nodeID string) (string, error)
 }
+
+// convertZones converts existing zone labels to NodeAffinity.
+func convertZones(pv *v1.PersistentVolume, topologyKey string) error {
+	zones := []string{}
+	for k, v := range pv.ObjectMeta.Labels {
+		if k == v1.LabelZoneFailureDomain {
+			zones = append(zones, v)
+		}
+	}
+
+	// If there are no zone labels, skip
+	if len(zones) == 0 {
+		return nil
+	}
+
+	// Make sure the necessary fields exist
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
+
+	hasTopology := func(pv *v1.PersistentVolume) bool {
+		for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+			for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+				if r.Key == topologyKey {
```

Comment:
Hm, IMO we should skip (don't convert zones to node affinity) if there is **any** node affinity set (be it in-tree or CSI).

Personally I'd like the scope of this patch to be: convert zones to topology when possible, but don't overwrite object fields when they have data. It might make the troubleshooting more difficult. @msau42, what do you think?

PS: You owe me a dumb question now.

Reply:
> Hm, IMO we should skip (don't convert zones to node affinity) if there is any node affinity set (be it in-tree or CSI).

This seems reasonable to me. Though I feel like that's not what the code is currently doing (I might just be confused)?

Reply:
Today, there are two in-tree cases to consider for migration:

1) PV only has zone labels on it
2) PV has zone labels + PV.NodeAffinity set (with the beta labels).

We need to handle both cases for migration because support for the old beta labels will be going away.

Reply:
Added logic to handle both cases, PTAL.